overlay.go

```diff
@@ -0,0 +1,346 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package charm
+
+import (
+	"bytes"
+	"encoding/gob"
+	"fmt"
+	"math"
+	"reflect"
+	"strings"
+)
+
+// ExtractStaticAndOverlayParts splits the bundle data into a static/shared and
+// overlay-specific bundle so that their union yields bd. To decide whether a
+// field is overlay-specific, the implementation uses reflection and
+// recursively scans the BundleData fields looking for fields annotated with
+// the "overlay-only: true" tag.
+//
+// To produce the static/shared bundle, the original bundle is filtered and all
+// overlay-specific values are set to the zero value for their type. To produce
+// the overlay-specific bundle, we once again filter the original bundle but
+// this time zero out fields that do not contain any descendant fields that are
+// overlay-specific.
+//
+// To clarify how this method works let's consider a bundle created via the
+// yaml blob below:
+//
+//   applications:
+//     apache2:
+//       charm: cs:apache2-26
+//       offers:
+//         my-offer:
+//           endpoints:
+//           - apache-website
+//           - website-cache
+//         my-other-offer:
+//           endpoints:
+//           - apache-website
+//   series: bionic
+//
+// The "offers" and "endpoints" attributes are overlay-specific fields. If we
+// were to run this method and then marshal the results back to yaml we would
+// get:
+//
+// The static/shared bundle:
+//
+//   applications:
+//     apache2:
+//       charm: cs:apache2-26
+//   series: bionic
+//
+// The overlay-specific bundle:
+//
+//   applications:
+//     apache2:
+//       offers:
+//         my-offer:
+//           endpoints:
+//           - apache-website
+//           - website-cache
+//         my-other-offer:
+//           endpoints:
+//           - apache-website
+//
+// The two bundles returned by this method are copies of the original bundle
+// data and can thus be safely manipulated by the caller.
+func ExtractStaticAndOverlayParts(bd *BundleData) (static, overlay *BundleData, err error) {
+	if static, err = cloneBundleData(bd); err != nil {
+		return nil, nil, err
+	}
+
+	if overlay, err = cloneBundleData(bd); err != nil {
+		return nil, nil, err
+	}
+
+	_ = visitField(&visitorContext{structVisitor: clearOverlayFields}, static)
+	_ = visitField(&visitorContext{structVisitor: clearNonOverlayFields}, overlay)
+	return static, overlay, nil
+}
+
+// cloneBundleData uses the gob package to perform a deep copy of bd.
+func cloneBundleData(bd *BundleData) (*BundleData, error) {
+	var buf bytes.Buffer
```

Comment:
My decision was purely to avoid importing any additional dependencies into the charm package. I could have gone with `bundle -> yaml -> bundle` (or json) instead but I thought using gob would be a bit more generic.

In terms of performance, bundles are not that large so I don't anticipate this being an issue. If you think this might be an issue I can swap the implementation to use json instead. Thoughts?

Reply:
I forgot to mention another reason for picking gob. With the gob approach you get a copy of all public fields whereas with json or yaml you might lose a field if someone adds a `json:"-"` annotation for a field that should not be exported etc.

Reply:
fwiw I prefer the copy-by-reflection (rog-go/deepcopy and similar), or just a bunch of nested "foo = *foo" sort of copies. Its all just a memory copy anyway, serializing and deserializing seems rather inefficient to just get a memory copy.

I don't expect it to be an actual issue, just a preference.